pennylane/ops/qubit/matrix_ops.py

```diff
@@ -0,0 +1,483 @@
+        # to the left and right with the correct amount of identity blocks.
+
+        self._padding_left = control_int * len(U)
+        self._padding_right = 2 ** len(wires) - len(U) - self._padding_left
+        self._CU = None
+
+        super().__init__(*params, wires=wires, do_queue=do_queue)
+
+    def _matrix(self, *params):
+        if self._CU is None:
+            self._CU = block_diag(np.eye(self._padding_left), self.U, np.eye(self._padding_right))
+
+        params = list(params)
+        params[0] = self._CU
+        return super()._matrix(*params)
+
+    @property
+    def control_wires(self):
+        return self._control_wires
+
+    @staticmethod
+    def _parse_control_values(control_wires, control_values):
+        """Ensure any user-specified control strings have the right format."""
+        if isinstance(control_values, str):
+            if len(control_values) != len(control_wires):
+                raise ValueError("Length of control bit string must equal number of control wires.")
+
+            # Make sure all values are either 0 or 1
+            if any(x not in ["0", "1"] for x in control_values):
+                raise ValueError("String of control values can contain only '0' or '1'.")
+
+            control_int = int(control_values, 2)
+        else:
+            raise ValueError("Alternative control values must be passed as a binary string.")
+
+        return control_int
+
+    def _controlled(self, wire):
+        ctrl_wires = sorted(self.control_wires + wire)
+        ControlledQubitUnitary(*self.parameters, control_wires=ctrl_wires, wires=self._target_wires)
+
+
+class DiagonalQubitUnitary(DiagonalOperation):
+    r"""DiagonalQubitUnitary(D, wires)
+    Apply an arbitrary fixed diagonal unitary matrix.
+
+    **Details:**
+
+    * Number of wires: Any (the operation can act on any number of wires)
+    * Number of parameters: 1
+    * Gradient recipe: None
+
+    Args:
+        D (array[complex]): diagonal of unitary matrix
+        wires (Sequence[int] or int): the wire(s) the operation acts on
+    """
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+    grad_method = None
+
+    @classmethod
+    def _eigvals(cls, *params):
+        D = np.asarray(params[0])
+
+        if not np.allclose(D * D.conj(), np.ones_like(D)):
+            raise ValueError("Operator must be unitary.")
+
+        return D
+
+    @staticmethod
+    def decomposition(D, wires):
+        return [QubitUnitary(np.diag(D), wires=wires)]
+
+    def adjoint(self):
+        return DiagonalQubitUnitary(qml.math.conj(self.parameters[0]), wires=self.wires)
+
+    def _controlled(self, control):
+        DiagonalQubitUnitary(
+            qml.math.concatenate([np.array([1, 1]), self.parameters[0]]),
+            wires=Wires(control) + self.wires,
+        )
+
+
+# When this gate no longer depends on ``ControlledQubitUnitary
+# please move it to ``non_parametric_ops.py``
+
+
+class MultiControlledX(ControlledQubitUnitary):
```

Comment:
Hi @albi3ro, will be going for a full review. One thing before that is that it would be great to ensure that the reorganization is with respect to how the operations behave, not how they are coded. E.g., here `MultiControlledX` should probably live in `non_parametric_ops` rather than this file.

Reply:
An external contributor is editing 'MultiControlledX' to no longer inherit from 'ControlledQubitUnitary', at which point it can be moved to non-parametric ops.

Reply:
@albi3ro I would strongly recommend organizing the code around how the operation behaves/its category, rather than around how it is coded up 🤔 

Even though `MultiControlledX` is a matrix op internally, I feel like this is an implementation detail? How an operation is implemented should ideally not require it to change files.